flashinfer/gemm.py

```diff
@@ -1240,3 +1249,174 @@ def group_deepgemm_fp8_nt_groupwise(
+    )
+    b_cudnn_tensor = graph.tensor(
+        name="b", dim=b_shape, stride=b_stride, data_type=ab_type
+    )
+    scale_cudnn_tensor = graph.tensor(
+        name="scale", dim=(1, 1, 1), stride=(1, 1, 1), data_type=cudnn.data_type.FLOAT
+    )
+    c_cudnn_tensor = graph.matmul(
+        name="matmul",
+        A=a_cudnn_tensor,
+        B=b_cudnn_tensor,
+        compute_data_type=cudnn.data_type.FLOAT,
+    )
+    c_cudnn_tensor.set_name("c").set_data_type(cudnn.data_type.FLOAT)
+    c_final_cudnn_tensor = graph.mul(
+        name="scale_mul",
+        a=c_cudnn_tensor,
+        b=scale_cudnn_tensor,
+        compute_data_type=cudnn.data_type.FLOAT,
+    )
+    c_final_cudnn_tensor.set_name("c_final").set_output(True).set_data_type(o_type)
+
+    a_cudnn_tensor.set_uid(UIDs.A_UID.value)
+    b_cudnn_tensor.set_uid(UIDs.B_UID.value)
+    scale_cudnn_tensor.set_uid(UIDs.SCALE_UID.value)
+    c_final_cudnn_tensor.set_uid(UIDs.O_UID.value)
+
+    graph.validate()
+    graph.build_operation_graph()
+    graph.create_execution_plans([cudnn.heur_mode.A, cudnn.heur_mode.FALLBACK])
+    graph.check_support()
+    graph.build_plans()
+
+    return graph
+
+
+def execute_gemm_with_per_tensor_q_graph(graph, a, b, scale_tensor, c_final):
+    variant_pack = {
+        UIDs.A_UID.value: a,
+        UIDs.B_UID.value: b,
+        UIDs.SCALE_UID.value: scale_tensor,
+        UIDs.O_UID.value: c_final,
+    }
+
+    workspace = torch.empty(
+        graph.get_workspace_size(), device="cuda", dtype=torch.uint8
+    )
+
+    graph.execute(variant_pack, workspace)
+
+
+def gemm_f8f8_f32_bf16(
+    a: torch.Tensor,  # (batch, m, k)
+    b: torch.Tensor,  # (batch, k, n)
+    dq_scale: torch.Tensor,  # (1, 1, 1)
+    out: Optional[torch.Tensor] = None,  # (batch, m, n)
+):
+    """Perform FP8 GEMM with per-tensor dequantization and BF16 output.
+
+    This function uses cached cuDNN graphs for optimal performance.
+
+    Args:
+        a: FP8 input tensor A with shape (batch, m, k)
+        b: FP8 input tensor B with shape (batch, k, n)
+        dq_scale: Dequantization scale tensor with shape (1, 1, 1)
+        out: Optional output tensor with shape (batch, m, n). If None, will be allocated.
+
+    Returns:
+        BF16 output tensor with shape (batch, m, n)
+    """
+    _check_cudnn_availability()
+
+    if out is None:
+        out = torch.empty(
+            a.shape[0], a.shape[1], b.shape[2], dtype=torch.bfloat16, device=a.device
+        )
+
+    graph = build_gemm_with_per_tensor_q_graph(
+        a.shape,
+        a.stride(),
+        b.shape,
+        b.stride(),
+        cudnn.data_type.FP8_E4M3,
+        cudnn.data_type.BFLOAT16,
+    )
+
+    execute_gemm_with_per_tensor_q_graph(graph, a, b, dq_scale, out)
+    return out
+
+
+def gemm_f8f8_f32_fp16(
+    a: torch.Tensor,  # (batch, m, k)
+    b: torch.Tensor,  # (batch, k, n)
+    dq_scale: torch.Tensor,  # (1, 1, 1)
+    out: Optional[torch.Tensor] = None,  # (batch, m, n)
```

Comment:
Please also add an `out_dtype` argument, in case `out` is not provided, we create temporary output buffer with `out_dtype`.
`out_dtype` should default to `torch.bfloat16` (which is used more frequently than fp16).

Reply:
Thank you for comment.

In this PR I added 2 apis: `gemm_f8f8_f32_bf16` and `gemm_f8f8_f32_fp16`. And you prefer to use single api:
`gemm_fp8` with the output_dtype as the argument in it.

and in the next PR we can add gemm_fp8_block_scale. Is my understanding correct?

Reply:
btw there is an early api: https://github.com/flashinfer-ai/flashinfer/blob/f153369384db9298f52b00d8e0fc53cb0d1f0f49/flashinfer/gemm.py#L712-L779 that calls into cublas bmm fp8, is this possible to unify them?

Reply:
In the latest commit, now we merge into single `bmm_fp8` and add a new `backend` argument, please help to check if this align with you? thanks!

Reply:
I'm not pretty sure if their semantics are equivalent as the cublas API is batched matmul. Can you double check?

Reply:
I checked the spec

for [cublas](https://docs.nvidia.com/cuda/cublas/#id5), only `m,n,k` is supported in `cublasSgemm`; there is extra `batchCount` in `cublasSgemmBatched` ;

for [torch.matmul](https://docs.pytorch.org/docs/stable/generated/torch.matmul.html), If both arguments are at least 1-dimensional and at least one argument is N-dimensional (where N > 2), then a batched matrix multiply is returned.

for cudnn, batch gemm and gemm are supported is single API [cudnn.matmul](https://docs.nvidia.com/deeplearning/cudnn/frontend/latest/operations/Matmul.html)


Also in the test, we tested both bs=1 and bs>1 case.

Reply:
sounds great!